### ISSUE TITLE
histogram_loss implementation?

### ISSUE THREAD

**Issue by sumo8291** [valerystrizh/pytorch-histogram-loss#6]
In HistogramLoss of "loss.py", if you do torch.sum(histogram_neg); I don't get the sum to be 1. histogram_neg is a Probability Density function, so it should sum to 1. Moreover, histogram_pos_cdf should have atleast one element equal to 1, as it is a CDF. However it is not the case in your code. Are you sure your code is correct or am I missing something?

**Comment by valerystrizh** [comment]
You are right, histogram_pos and histogram_neg were divided by 2 and each of their sum equaled to 0.5. This bug is fixed and now these PMFs sum up to  1. Confusing variable histogram_pos_cdf is renamed to histogram_pos_cdf_by_step where the last term is equal to 1 as you expect

**Comment by sumo8291** [comment]
I still don't think it has been resolved properly. Why are you dividing the pos_size and neg_size by 2? You are anyways taking the upper diagonal elements of the classes_eq for calculation s_inds. So it should anyways contain the right number of positive and negative pairs.
![new](https://user-images.githubusercontent.com/8149573/49443761-5d6f8900-f821-11e8-9e76-9b05ec584c32.png)

**Comment by valerystrizh** [comment]
You were quite right, there was a mistake in histogram method. Please, check if it is all right now

**Comment by sumo8291** [comment]
I am again sorry to say that it is not correct.
The histogram_pos_cdf_by_step has negative values.
![new](https://user-images.githubusercontent.com/8149573/49473305-cb3da400-f865-11e8-89da-93d8a9047b62.png)

**Comment by valerystrizh** [comment]
There is the next version, but it still does not work with bad number of bins, e.g. which are not ended by 1

**Comment by sumo8291** [comment]
Hi,
In your new implementation,  how did you guarantee that the weights chosen are assigned to two adjacent nodes as in equation (2) of the paper? 

Thanks

**Comment by valerystrizh** [comment]
Hi,
The implementation of choosing weights which are assigned to two adjacent nodes is the same as it was before your first issue. The old implementation is https://github.com/valerystrizh/pytorch-histogram-loss/blob/613b5afd46d6680cfd0e1d50ea3d366f4838e64f/losses.py
What is changed is the comparison of floats and checking whether some values disappear with an inappropriate number of bins. Unfortunately, the implementation still requires choosing this parameter. This is going to be fixed in the nearest future.

The guarantee itself is analytical.

**Comment by sumo8291** [comment]
thanks a lot
